### ISSUE TITLE
I can not restore dependencies from package-lock.json

### ISSUE THREAD

**Issue by cawa-93** [npm/npm#17091]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [x] npm is producing an incorrect install.
  - [x] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
I have a `package.json`. After install dependencies (`npm i`), a `package-lock.json` was generated.
If you use this file to restore the dependency tree then the tree will be broken

#### How can the CLI team reproduce the problem?
First clone my package.json without package-lock.json
```
$ git clone git@github.com:cawa-93/npm-issue.git
$ cd npm-issue
$ git checkout 80eb3bb5d81c8d69a3de9f87fe7d72f9cb6fd4a5
```
Install dependency and check `keygrip` package
```
$ npm i && npm ls keygrip
npm notice created a lockfile as package-lock.json. You should commit this file.
added 1267 packages in 213.16s
madpost@1.0.0 /run/media/alex/LinWin/Develop/npm-issue
└─┬ adonis-framework@3.0.13
  └─┬ node-cookie@1.0.4
    └── keygrip@1.0.1  (git://github.com/poppinss/keygrip.git#f7b41c2f305cfa7030d7da9e64af9790aa3be93a)
```
Delete all node_modules, install dependency from package-lock.json and check `keygrip` package again
```
$ rm -rf ./node_modules && npm i && npm ls keygrip
added 1263 packages in 149.225s
madpost@1.0.0 /run/media/alex/LinWin/Develop/npm-issue
└─┬ adonis-framework@3.0.13
  └─┬ node-cookie@1.0.4
    └── UNMET DEPENDENCY keygrip@git://github.com/poppinss/keygrip.git#develop

npm ERR! missing: keygrip@git://github.com/poppinss/keygrip.git#develop, required by node-cookie@1.0.4
```
After restore, the `package-lock.json` also changed: https://github.com/cawa-93/npm-issue/commit/74d29a9ec4f80e3a4ab250db552585599be3f840
### supporting information:

 - `npm -v` prints: 5.0.3
 - `node -v` prints: v8.0.0
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: Linux

**Comment by xgqfrms-GitHub** [comment]
https://github.com/npm/npm/releases/tag/v5.0.0

**Comment by cawa-93** [comment]
@xgqfrms-GitHub What are you trying to say?

**Comment by diversario** [comment]
Yep, I just attempted to `npm install` from a shrinkwrap and at least one dependency is missing two `git` dependencies.

**Comment by arndtjenssen** [comment]
Same problem here. Doing an npm install with a package-lock.json updated from a co-worker via git does result in missing dependencies, even after a complete rebuild..

**Comment by cawa-93** [comment]
Any comment from the @npm members?

**Comment by AtiX** [comment]
In my case (see https://github.com/npm/npm/issues/17252), I noticed that the depdency was sorted into the wrong (I assume) hierarchy level in package-lock.json, this may be the case here as well?

**Comment by carnesen** [comment]
We ran into this bug too. For what it's worth, in our case the unstable dependency is a `git+ssh://git@github.com/...` one, and I also noticed that it appears to be at the wrong level of the hierarchy in package-lock.json.

**Comment by DominicBoettger** [comment]
Some packages are missing in package-lock.json. In my case for example body-parser which is needed by a submodule. It works when installing without package-lock. I get the error as soon as the package-lock.json exists.

**Comment by rjmount** [comment]
As part of my build processes, the only consistent fix for this that I have found is to delete the package-lock.json before every npm install/update. The alternative is to narrow down all the offending dependencies and explicitly define them in my original package.json myself.

I really don't feel like I should need to do either of these things, but they are the only solutions I've found.

**Comment by cawa-93** [comment]
Today check this issue with npm **v5.0.4** and node **v8.1.3**:
```
$ npm i && npm ls keygrip
npm notice created a lockfile as package-lock.json. You should commit this file.
added 1303 packages in 166.392s
madpost@1.0.0 /run/media/alex/LinWin/Develop/npm-issue
└─┬ adonis-framework@3.0.14
  └─┬ node-cookie@1.0.4
    └── keygrip@1.0.1  (git://github.com/poppinss/keygrip.git#f7b41c2f305cfa7030d7da9e64af9790aa3be93a)
```
```
$ rm -rf ./node_modules && npm i && npm ls keygrip
added 1299 packages in 91.354s
madpost@1.0.0 /run/media/alex/LinWin/Develop/npm-issue
└─┬ adonis-framework@3.0.14
  └─┬ node-cookie@1.0.4
    └── UNMET DEPENDENCY keygrip@git://github.com/poppinss/keygrip.git#develop

npm ERR! missing: keygrip@git://github.com/poppinss/keygrip.git#develop, required by node-cookie@1.0.4
```
As before, not all packages was installed (1303 ≠ 1299)

**Comment by dman777** [comment]
I hit this issue also with npm ver 5.0.3 and node 8.1.2. I kept getting errors of missing dependencies when trying to run gulp. Only when I removed package-lock.json and re-installed my node_modules did it fix the issue.

**Comment by kivervinicius** [comment]
+1

**Comment by carnesen** [comment]
This looks very similar to #16839 which is fixed in npm@5.1 released a couple days ago. As of today (July 7), the Node.js "current" release (8.1.3) still comes with npm@5.0, but you can `npm install -g npm@5.1`. In my project that was previously suffering from package-lock.json-related issues, I find everything works as expected with npm@5.1.

**Comment by cawa-93** [comment]
Today I tested it on **v5.2.0**
When I first installed, I received a message:
```
$ npm i
added 1506 packages in 167.012s
```
At the second:
```
$ rm -rf ./node_modules && npm i
added 1391 packages in 60.074s
```
Why did you install less packages when reinstalling?

**Comment by yosuke-furukawa** [comment]
I hit same issue. I don't understand the reason, but I always put `--no-save` option when execute `npm install`. so I can avoid to update `package-lock.json` unintentionally.

**Comment by fjferrer** [comment]
Similar issue here with 'file:' style dependencies (See #17527)

My environment:
Node version: 6.11.1
NPM version: 5.3.0

[Here](https://gist.github.com/fjferrer/2bdd6aa620d37f5fc31ff1334acf5320) is a gist to reproduce it.

**Comment by arndtjenssen** [comment]
until this is fixed just a hint. one can disable package-lock.json completely by adding:
`package-lock = false`
to .npmrc

This works well for us...

**Comment by kanreisa** [comment]
I had same issue. `package-lock.json` (and `npm@5`) is broken! 😡😡💢 

> `package-lock = false`

thx workaround.

```
Windows 10 RS2, Azure App Services (Windows Server)
node: v8.1.4
npm: 5.3.0
```

**Comment by cchamberlain** [comment]
We've been trying to use `package-lock.json` from the start but have been seeing inconsistency. We have a script that deletes `node_modules` and runs `npm i` and somehow that causes `package-lock.json` churn, mostly around `"requires": true` nodes being added and removed.

My understanding was that an `npm install` will not regenerate `package-lock.json` if it exists, but rather, install what's declared in the existing lock - @iarna is this correct? Is the `package-lock.json` churn suggestive that some check is being missed? Is the `package-lock.json` output non-deterministic across different npm 5 versions?

Happy to dig and put together a PR if you can give any clues as to where the problem may lie.

**Comment by cchamberlain** [comment]
This is a pretty serious issue affecting everyone in my org. Would be nice to get some sort of update or nudge in the right direction to get working on a PR. After wiping out my `package-lock.json` and `node_modules` and doing a fresh install, I end up with no `webpack` and no errors 😿.

UPDATE: `webpack` does appear to be installed, but `npm ls webpack` doesn't pick up on it.

**Comment by seepel** [comment]
I've found that a temporary work around is to explicitly install the offending dependency. For example, my issue was with node-soap as a sub dependency of node-ews, so I setup a shell script that re-installs it.
```bash
npm install
npm install soap@https://github.com/cumberlandgroup/node-soap/tarball/master
```
This way, the majority of dependent packages are still locked, and the only indeterministic packages are node-soap, and it's sub dependencies.

**Comment by eljefedelrodeodeljefe** [comment]
ping @ceejbot (maybe?). Running into the same issue.

**Comment by kennethaasan** [comment]
I have the same issue. We have some private dependencies which is packed in tgz which is "missing" with npm ls after installing with package-lock.json. If there is no package-lock.json npm ls works fine.

**Comment by senseysensor** [comment]
the same here. :(

**Comment by ghost** [comment]
Exact same problem here.
npm - 5.6.0
node - 8.7.0

When I install using npm install I can no longer depend on all my dependencies being installed.  This is now also affecting our auto deployment system and is incredibly frustrating.  I hope the devs here are able to fix the growing issues with npm@5.x soon or, at least, provide some insight as to what is being done to address them.

**Comment by CCharlieLi** [comment]
Same here. It's a private package installed from GitHub private repo, which is not showing in package-lock.json.(I can find it in yarn.lock btw) This really prevent me using `npm ci` as it keeps saying `npm ERR! Missing: XXX`.

npm - 5.7.1
node - v8.9.0
